diff -urN nesterdc-5.0-pre4/README nesterdc-5.0-pre4-fix/README
--- nesterdc-5.0-pre4/README	Wed Jan 23 17:09:01 2002
+++ nesterdc-5.0-pre4-fix/README	Fri Jan 25 02:08:49 2002
@@ -12,6 +12,7 @@
          If you set DC pad on "port 0, 2", it's recognized as "NES pad 0, 1", 
          not "NES pad 0, 2". 
  * speedup. (works almost 60fps). 
+ * fullscreen support 
 
 # 4.0
  * cpu/ppu/apu engine was ported from 
Binary files nesterdc-5.0-pre4/bzip2-1.0.1/libbz2.a and nesterdc-5.0-pre4-fix/bzip2-1.0.1/libbz2.a differ
diff -urN nesterdc-5.0-pre4/dcmake nesterdc-5.0-pre4-fix/dcmake
--- nesterdc-5.0-pre4/dcmake	Wed Jan 23 17:27:59 2002
+++ nesterdc-5.0-pre4-fix/dcmake	Fri Jan 25 03:06:54 2002
@@ -2,7 +2,7 @@
 
 #============================================================
 # Edit for your environment
-export NDC_BASE=/home/tekezo/work/tmp/nesterdc-5.0-pre4
+export NDC_BASE=/home/tekezo/work/tmp/nesterdc-5.0-pre4-fix
 export CROSS_BASE=/home/const/cross/dreamcast
 
 #NESTER_OPTIONS="${NESTER_OPTIONS} -D__PRECISION__EMULATE__"
diff -urN nesterdc-5.0-pre4/kos-1.1.6/kernel/arch/dreamcast/hardware/ta/ta.c nesterdc-5.0-pre4-fix/kos-1.1.6/kernel/arch/dreamcast/hardware/ta/ta.c
--- nesterdc-5.0-pre4/kos-1.1.6/kernel/arch/dreamcast/hardware/ta/ta.c	Wed Jan 23 17:09:01 2002
+++ nesterdc-5.0-pre4-fix/kos-1.1.6/kernel/arch/dreamcast/hardware/ta/ta.c	Fri Jan 25 02:08:49 2002
@@ -13,7 +13,7 @@
 #include <arch/irq.h>
 #include <kos/thread.h>
 
-CVSID("ta.c,v 1.3 2002/01/23 08:09:01 tekezo Exp");
+CVSID("ta.c,v 1.4 2002/01/24 17:08:49 tekezo Exp");
 
 /*
 
@@ -109,8 +109,8 @@
 
 /* TA lists that you can activate */
 #define TA_LIST_OPAQUE_POLYS	1
-#define TA_LIST_TRANS_POLYS	2
-#define TA_LIST_OPAQUE_MODS	4
+#define TA_LIST_OPAQUE_MODS	2
+#define TA_LIST_TRANS_POLYS	4
 #define TA_LIST_TRANS_MODS	8
 #define TA_LIST_PUNCH_THRU	16
 
@@ -867,9 +867,7 @@
 	    && (list_complete == ta_state.lists)
 	    && render_complete) {
 		static int screen_index = 0;
-		/* Switch view address to the "good" buffer */
-		
-#if 0
+#if 1
 		if (!(dc_ta_enqueue (ta_screen_addrs[screen_index])))
 			return;
 #else
Binary files nesterdc-5.0-pre4/kos-1.1.6/kernel/arch/dreamcast/hardware/ta/ta.o and nesterdc-5.0-pre4-fix/kos-1.1.6/kernel/arch/dreamcast/hardware/ta/ta.o differ
Binary files nesterdc-5.0-pre4/kos-1.1.6/kernel/build/banner.o and nesterdc-5.0-pre4-fix/kos-1.1.6/kernel/build/banner.o differ
Binary files nesterdc-5.0-pre4/kos-1.1.6/kernel/build/ta.o and nesterdc-5.0-pre4-fix/kos-1.1.6/kernel/build/ta.o differ
Binary files nesterdc-5.0-pre4/kos-1.1.6/lib/libkallisti.a and nesterdc-5.0-pre4-fix/kos-1.1.6/lib/libkallisti.a differ
diff -urN nesterdc-5.0-pre4/nester/debug/debug.h nesterdc-5.0-pre4-fix/nester/debug/debug.h
--- nesterdc-5.0-pre4/nester/debug/debug.h	Sun Jan  6 16:31:28 2002
+++ nesterdc-5.0-pre4-fix/nester/debug/debug.h	Fri Jan 25 02:08:49 2002
@@ -30,8 +30,12 @@
 #  define LOGHEX(X) printf("%x\n", X);
 #  define ERROR(MSG) fprintf(stderr, "[ERROR] %s\n", MSG);
 #  define CRASH_POINT()
-#  include <assert.h>
-#  define ASSERT(EXPR) assert(EXPR)
+#  if 0
+#    include <assert.h>
+#    define ASSERT(EXPR) assert(EXPR)
+#  else
+#    define ASSERT(EXPR)
+#  endif
 #else
 #  include "dc_utils.h"
 #  define LOG(MSG)
diff -urN nesterdc-5.0-pre4/nester/dreamcast/dc_main.cpp nesterdc-5.0-pre4-fix/nester/dreamcast/dc_main.cpp
--- nesterdc-5.0-pre4/nester/dreamcast/dc_main.cpp	Wed Jan 23 17:09:01 2002
+++ nesterdc-5.0-pre4-fix/nester/dreamcast/dc_main.cpp	Fri Jan 25 03:06:01 2002
@@ -858,7 +858,7 @@
   vmu_icon_draw_string (vmu_screen, 3, 7, str);
   
   vmu_icon_draw_string (vmu_screen, 1, 20, "NESTERDC");
-  vmu_icon_draw_string (vmu_screen, 3, 26, "5.0-PRE4");
+  vmu_icon_draw_string (vmu_screen, 3, 26, "5.0-PRE4+");
   
   timer_spin_sleep (50); /* need wait */
   vmu_icon_flip (vmu_screen, mldc);
@@ -945,7 +945,7 @@
   load_bmp (options_image, "/cd/pics/options.bmp");
   
   display_rawimage (startup_image);
-  draw_string(10, 210, _white, _black, "NesterDC 5.0-pre4");
+  draw_string(10, 210, _white, _black, "NesterDC 5.0-pre4-fix");
   dc_vid_flip_fill_renderer (draw_type_fullscreen);
   
   init_menus ();
Binary files nesterdc-5.0-pre4/nester/dreamcast/dc_main.o and nesterdc-5.0-pre4-fix/nester/dreamcast/dc_main.o differ
Binary files nesterdc-5.0-pre4/nester/dreamcast/dc_sound.o and nesterdc-5.0-pre4-fix/nester/dreamcast/dc_sound.o differ
diff -urN nesterdc-5.0-pre4/nester/dreamcast/dc_utils.c nesterdc-5.0-pre4-fix/nester/dreamcast/dc_utils.c
--- nesterdc-5.0-pre4/nester/dreamcast/dc_utils.c	Wed Jan 23 17:09:01 2002
+++ nesterdc-5.0-pre4-fix/nester/dreamcast/dc_utils.c	Fri Jan 25 02:56:41 2002
@@ -78,10 +78,6 @@
   float u;
   float x1, x2;
   
-  ta_begin_render();
-  ta_poly_hdr_txr(&poly, TA_OPAQUE, TA_RGB565, 
-                  512, 256, textureaddr, TA_NO_FILTER);
-  
   if (dt == draw_type_fullscreen)
   {
     u = 320.0/512;
@@ -101,6 +97,9 @@
     x2 = 576;
   }
   
+  ta_begin_render();
+  ta_poly_hdr_txr(&poly, TA_OPAQUE, TA_RGB565, 
+                  512, 256, textureaddr, TA_NO_FILTER);
   ta_commit_poly_hdr(&poly);
   
   vert.flags = TA_VERTEX_NORMAL;
@@ -151,9 +150,7 @@
   int i;
   
   for (i = 0; i < DC_TEXTURE_ADDRS_SIZE; ++i)
-  {
     dc_texture_addrs[i] = ta_txr_allocate (512 * 256 * 2);
-  }
   
   dc_vid_clear ();
   
@@ -199,9 +196,6 @@
   static int texture_index = 0;
   int i;
   uint16 *d, *s;
-
-  while (!(ta_render_done ()))
-    ;
   
   d = ta_txr_map (dc_texture_addrs[texture_index]);
   s = dc_screen;
@@ -218,6 +212,7 @@
   }
   else
   {
+    d += 256 * 2 * 8;
     s += 32;
     for (i = 0; i < 224; ++i)
     {
@@ -227,6 +222,9 @@
     }
     dc_wait_sq_cpy_done ();
   }
+  
+  while (!(ta_render_done ()))
+    ;
   
   draw_screen (dc_texture_addrs[texture_index], dt);
   texture_index = (texture_index + 1) & 0x01;
Binary files nesterdc-5.0-pre4/nester/dreamcast/dc_utils.o and nesterdc-5.0-pre4-fix/nester/dreamcast/dc_utils.o differ
diff -urN nesterdc-5.0-pre4/nester/dreamcast/sound/aica_fw.h nesterdc-5.0-pre4-fix/nester/dreamcast/sound/aica_fw.h
--- nesterdc-5.0-pre4/nester/dreamcast/sound/aica_fw.h	Mon Dec 24 00:15:59 2001
+++ nesterdc-5.0-pre4-fix/nester/dreamcast/sound/aica_fw.h	Fri Jan 25 02:08:49 2002
@@ -10,7 +10,7 @@
 149, 229, 4, 0, 80, 225, 51, 0, 0, 10, 16, 17, 159, 229, 0, 0, 145, 229, 0, 32, 144, 229, 8, 0, 82, 227, 44, 0, 0, 10, 
 16, 0, 82, 227, 1, 0, 0, 10, 0, 0, 160, 227, 112, 168, 27, 233, 0, 64, 160, 227, 232, 0, 159, 229, 232, 48, 159, 229, 0, 96, 
 144, 229, 0, 80, 147, 229, 0, 32, 150, 229, 0, 0, 149, 229, 162, 17, 160, 225, 33, 1, 0, 235, 0, 0, 141, 229, 0, 0, 160, 227, 
-1, 80, 160, 227, 4, 32, 160, 225, 192, 16, 159, 229, 0, 224, 145, 229, 0, 192, 158, 229, 17, 26, 160, 227, 0, 48, 160, 225, 200, 96, 
+1, 80, 160, 227, 4, 32, 160, 225, 192, 16, 159, 229, 0, 224, 145, 229, 0, 192, 158, 229, 17, 26, 160, 227, 0, 48, 160, 225, 220, 96, 
 160, 227, 128, 64, 160, 227, 4, 192, 141, 229, 8, 96, 141, 229, 2, 101, 160, 227, 16, 80, 141, 229, 12, 64, 141, 229, 81, 0, 0, 235, 
 116, 192, 159, 229, 0, 224, 156, 229, 0, 80, 142, 229, 10, 11, 134, 226, 128, 80, 159, 229, 128, 16, 159, 229, 0, 192, 149, 229, 20, 224, 
 144, 229, 0, 224, 140, 229, 0, 64, 145, 229, 15, 32, 132, 226, 0, 80, 145, 229, 2, 0, 85, 225, 252, 255, 255, 218, 247, 255, 255, 234, 
diff -urN nesterdc-5.0-pre4/nester/dreamcast/sound/main.c nesterdc-5.0-pre4-fix/nester/dreamcast/sound/main.c
--- nesterdc-5.0-pre4/nester/dreamcast/sound/main.c	Mon Dec 24 00:15:59 2001
+++ nesterdc-5.0-pre4-fix/nester/dreamcast/sound/main.c	Fri Jan 25 02:08:49 2002
@@ -69,7 +69,7 @@
   }
   
   aica_play (0, BASE_ADDRESS, sound_bits, 0, *buf_len / (*bits / 8), 
-             *rate, 200, 128, 1);
+             *rate, 220, 128, 1);
   
   *play_start = 1;
   
Binary files nesterdc-5.0-pre4/nester/dreamcast/sound/prog.elf and nesterdc-5.0-pre4-fix/nester/dreamcast/sound/prog.elf differ
Binary files nesterdc-5.0-pre4/nester/dreamcast/sound/stream.drv and nesterdc-5.0-pre4-fix/nester/dreamcast/sound/stream.drv differ
diff -urN nesterdc-5.0-pre4/nester/nes/ppu/nes_ppu.cpp nesterdc-5.0-pre4-fix/nester/nes/ppu/nes_ppu.cpp
--- nesterdc-5.0-pre4/nester/nes/ppu/nes_ppu.cpp	Tue Jan 22 00:31:44 2002
+++ nesterdc-5.0-pre4-fix/nester/nes/ppu/nes_ppu.cpp	Fri Jan 25 02:10:59 2002
@@ -349,18 +349,9 @@
   current_frame_line++;
 }
 
+
 void NES_PPU::do_scanline_and_dont_draw()
 {
-  // mmc2 / punchout -- we must simulate the ppu for every line
-  if(parent_NES->ROM->get_mapper_num() == 9)
-  {
-#ifdef __PRECISION__EMULATE__
-    do_scanline_and_draw(dummy_buffer, 0);
-#else
-    do_scanline_and_draw(dummy_buffer);
-#endif
-  }
-  else
   // if sprite 0 flag not set and sprite 0 on current line
   if ((!sprite0_hit()) && 
       (current_frame_line >= ((uint32)(spr_ram[0]+1))) && 
@@ -378,12 +369,25 @@
     if(spr_enabled() || bg_enabled())
     {
       LOOPY_SCANLINE_START(loopy_v, loopy_t);
+      
+      /* mmc2 / punchout -- we must simulate the ppu for every line */
+      /* if skip, freeze when get password */
+      if(parent_NES->ROM->get_mapper_num() == 9)
+      {
+	if(bg_enabled())
+	  render_bg_skip_mapper9 ();
+	
+	if(spr_enabled())
+	  render_spr_skip_mapper9 ();
+      }
+      
       LOOPY_NEXT_LINE(loopy_v);
     }
     current_frame_line++;
   }
 }
 
+
 void NES_PPU::end_frame()
 {
 }
@@ -819,6 +823,45 @@
   }
 }
 
+
+/* for only mapper 9 (punchout) */
+void
+NES_PPU::render_bg_skip_mapper9()
+{
+  uint32 i;
+  
+  uint32 tile_x;
+  uint32 tile_y;
+  uint32 tile_y_bit;
+  uint32 name_addr;
+  
+  uint32 pattern_addr;
+  
+  tile_x = (loopy_v & 0x001F);
+  tile_y = (loopy_v & 0x03E0) >> 5;
+  tile_y_bit = (loopy_v & 0x7000) >> 12;
+  
+  name_addr = 0x2000 + (loopy_v & 0x0FFF);
+  
+  for(i = 33; i; i--)
+  {
+    pattern_addr = bg_pattern_table_addr + ((int32)VRAM(name_addr) << 4) + tile_y_bit;
+    CHECK_MMC2(pattern_addr);
+    
+    if ((tile_x & 0x1f) != 0x1f)
+    {
+      ++tile_x;
+      ++name_addr;
+    }
+    else
+    {
+      tile_x -= 0x1f;
+      name_addr ^= 0x041f; 
+    }
+  }
+}
+
+
 void NES_PPU::render_spr(uint16* buf)
 {
   int sprite_index;
@@ -1013,4 +1056,30 @@
 }
 
 
-
+/* for only mapper 9 (punchout) */
+void
+NES_PPU::render_spr_skip_mapper9()
+{
+  int sprite_index;
+  uint8* spr;
+  
+  int num_sprites;
+  
+  /* pre-calced sprite data */
+  sprites_per_line *spl = spr_per_line + current_frame_line;
+  uint8 *spr_index = spl->sprite_index;
+  
+  num_sprites = spl->num;
+  while (num_sprites--)
+  {
+    sprite_index = *spr_index++;
+    spr = spr_ram + (sprite_index * 4);
+    
+    CHECK_MMC2(spr[1] << 4);
+  }
+  
+  if (spl->num > 8)
+    LowRegs[2] |= 0x20;
+  else
+    LowRegs[2] &= 0xDF;
+}
diff -urN nesterdc-5.0-pre4/nester/nes/ppu/nes_ppu.h nesterdc-5.0-pre4-fix/nester/nes/ppu/nes_ppu.h
--- nesterdc-5.0-pre4/nester/nes/ppu/nes_ppu.h	Thu Jan 10 03:13:20 2002
+++ nesterdc-5.0-pre4-fix/nester/nes/ppu/nes_ppu.h	Fri Jan 25 02:08:49 2002
@@ -204,7 +204,10 @@
 #else
   void render_bg(uint16* buf);
 #endif
+  void render_bg_skip_mapper9 ();
+  
   void render_spr(uint16* buf);
+  void render_spr_skip_mapper9 ();
 };
 
 #endif
Binary files nesterdc-5.0-pre4/nester/nes/ppu/nes_ppu.o and nesterdc-5.0-pre4-fix/nester/nes/ppu/nes_ppu.o differ
Binary files nesterdc-5.0-pre4/nester/obj-dreamcast/dc_main.o and nesterdc-5.0-pre4-fix/nester/obj-dreamcast/dc_main.o differ
Binary files nesterdc-5.0-pre4/nester/obj-dreamcast/dc_sound.o and nesterdc-5.0-pre4-fix/nester/obj-dreamcast/dc_sound.o differ
Binary files nesterdc-5.0-pre4/nester/obj-dreamcast/dc_utils.o and nesterdc-5.0-pre4-fix/nester/obj-dreamcast/dc_utils.o differ
Binary files nesterdc-5.0-pre4/nester/obj-dreamcast/nes_ppu.o and nesterdc-5.0-pre4-fix/nester/obj-dreamcast/nes_ppu.o differ
diff -urN nesterdc-5.0-pre4/nester/sdl/sdlmake nesterdc-5.0-pre4-fix/nester/sdl/sdlmake
--- nesterdc-5.0-pre4/nester/sdl/sdlmake	Wed Jan 23 17:28:15 2002
+++ nesterdc-5.0-pre4-fix/nester/sdl/sdlmake	Fri Jan 25 03:33:40 2002
@@ -2,7 +2,7 @@
 
 #============================================================
 # Edit for your environment
-NESTER_BASE="/home/tekezo/work/tmp/nesterdc-5.0-pre4/nester"
+NESTER_BASE="/home/tekezo/work/tmp/nesterdc-5.0-pre4-fix/nester"
 SDL_CONFIG=sdl11-config 
 
 NESTER_OPTIONS="${NESTER_OPTIONS} -D__UNIX__"
Binary files nesterdc-5.0-pre4/obj/nester.bin and nesterdc-5.0-pre4-fix/obj/nester.bin differ
Binary files nesterdc-5.0-pre4/obj/nester.elf and nesterdc-5.0-pre4-fix/obj/nester.elf differ
diff -urN nesterdc-5.0-pre4/script/cdburn.sh nesterdc-5.0-pre4-fix/script/cdburn.sh
--- nesterdc-5.0-pre4/script/cdburn.sh	Sun Jan  6 16:28:45 2002
+++ nesterdc-5.0-pre4-fix/script/cdburn.sh	Fri Jan 25 02:08:49 2002
@@ -1,36 +1,37 @@
-#!/bin/sh
-
-# this copy script might help some people out. 
-# The information to create this came from here: http://mc.pp.se/dc/cdr.html
-# thanks Marcus
-#
-# I have a directory structure like this:
-#
-# /cdrtools (cdrecord, mkisofs, and IP.BIN are in here)
-# /cdrtools/files (GAMES directory, GFX directory, and 1ST_READ.BIN are in here)
-#
-# replace the dev=1,0,0 in the commands below with your burner device number, 
-# which is returned from the -scanbus command.
-#
-./cdrecord.exe -scanbus
-#
-# create an audio track with 4 seconds of silence 
-dd if=/dev/zero bs=2352 count=300 of=audio.raw
-#
-# up the speed to support your burner. It is set to 1 for safety :)
-# Not everyone has 16x burners.
-#
-./cdrecord dev=1,0,0 speed=1 -multi -audio audio.raw
-#
-# I get 0,11702 from -msinfo. You may get 0,11700. 
-# If so, fix the mkisofs command below.
-#
-./cdrecord dev=1,0,0 -msinfo
-#
-./mkisofs -r -C 0,11702 -o tmp.iso files
-#
-( cat IP.BIN ; dd if=tmp.iso bs=2048 skip=16 ) > data.raw
-#
-./cdrecord dev=1,0,0 speed=1 -multi -xa1 data.raw
-#
-# done
+#!/bin/sh
+
+# this copy script might help some people out. 
+# The information to create this came from here: http://mc.pp.se/dc/cdr.html
+# thanks Marcus
+#
+# I have a directory structure like this:
+#
+# /cdrtools (cdrecord, mkisofs, and IP.BIN are in here)
+# /cdrtools/files (GAMES directory, GFX directory, and 1ST_READ.BIN are in here)
+#
+# replace the dev=1,0,0 in the commands below with your burner device number, 
+# which is returned from the -scanbus command.
+#
+cdrecord.exe -scanbus
+#
+# create an audio track with 4 seconds of silence 
+dd if=/dev/zero bs=2352 count=300 of=audio.raw
+#
+# up the speed to support your burner. It is set to 1 for safety :)
+# Not everyone has 16x burners.
+#
+cdrecord dev=1,0,0 speed=1 -multi -audio audio.raw
+#
+# I get 0,11702 from -msinfo. You may get 0,11700. 
+# If so, fix the mkisofs command below.
+#
+cdrecord dev=1,0,0 -msinfo
+#
+mkisofs -r -C 0,11702 -o tmp.iso files
+#
+( cat IP.BIN ; dd if=tmp.iso bs=2048 skip=16 ) > data.raw
+#
+cdrecord dev=1,0,0 speed=1 -multi -xa1 data.raw
+#
+# done
+
